relentless/simulate/simulate.py

```diff
@@ -736,11 +755,100 @@ class InitializeRandomly(GenericOperation):
     masses : dict
         Masses of each particle type. Defaults to None, which means particles
         have unit mass.
+    diameters : dict
+        Diameter of each particle type. Defaults to None, which means particles
+        are randomly inserted without checking their sizes.
 
     """
-    def __init__(self, seed, N, V, T=None, masses=None):
+    def __init__(self, seed, N, V, T=None, masses=None, diameters=None):
         super().__init__(seed, N, V, T, masses)
 
+    @classmethod
+    def _make_orthorhombic(cls, V):
+        # get the orthorhombic bounding box
+        if isinstance(V, extent.TriclinicBox):
+            Lx,Ly,Lz,xy,xz,yz = V.as_array(extent.TriclinicBox.Convention.HOOMD)
+            aabb = numpy.array([Lx/numpy.sqrt(1.+xy**2+(xy*yz-xz)**2), Ly/numpy.sqrt(1.+yz**2), Lz])
+        elif isinstance(V, extent.ObliqueArea):
+            Lx,Ly,xy = V.as_array(extent.ObliqueArea.Convention.HOOMD)
+            aabb = numpy.array([Lx/numpy.sqrt(1.+xy**2), Ly])
+        else:
+            raise TypeError('Random initialization currently only supported in triclinic/oblique extents')
+        return aabb
+
+    @classmethod
+    def _random_particles(cls, seed, N, V):
+        rng = numpy.random.default_rng(seed)
+        aabb = cls._make_orthorhombic(V)
+
+        positions = aabb*rng.uniform(size=(sum(N.values()), len(aabb)))
+        positions = V.wrap(positions)
+
+        types = []
+        for i,Ni in N.items():
+            types.extend([i]*Ni)
+
+        return positions, types
+
+    @classmethod
+    def _pack_particles(cls, seed, N, V, diameters):
+        rng = numpy.random.default_rng(seed)
+        aabb = cls._make_orthorhombic(V)
+        dimension = len(aabb)
+        positions = numpy.zeros((sum(N.values()), dimension), dtype=numpy.float64)
+        types = []
+        trees = {}
+        Nadded = 0
+        # insert the particles, big to small
+        sorted_N = sorted(N.items(), key=lambda x : x[1], reverse=True)
+        for i,Ni in sorted_N:
+            # generate site coordinates, on orthorhombic lattices
+            di = diameters[i]
+            if dimension == 3:
+                # fcc lattice
+                a = numpy.sqrt(2.)*di
+                lattice = a*numpy.array([1.,1.,1.])
+                unitcell = a*numpy.array([[0.,0.,0.],[0.5,0.5,0.],[0.5,0.,0.5],[0.,0.5,0.5]])
+            elif dimension == 2:
+                a = di
+                b = numpy.sqrt(3.)*di
+                lattice = numpy.array([a,b])
+                unitcell = numpy.array([[0.,0.],[0.5*a,0.5*b]])
```

Comment:
I need to revisit these lattices and the shifting procedure to make sure I did it right.

Reply:
This is correct as written, so I am cleaning up and commenting.